### ISSUE TITLE
Why is there a significant drop in `val_ppl` after fixing data-loading bug?

### ISSUE THREAD

**Issue by bobqianic** [jzhang38/TinyLlama#77]
<img width="686" alt="image" src="https://github.com/jzhang38/TinyLlama/assets/129547291/2c050899-cad1-4628-9b6e-006e4ecdbd96">

**Comment by VatsaDev** [comment]
Hmm, thats really interesting. 

thinking about it, repeated data could have distorted the data patterns, making the model look at data in ways it shouldn't have, which could have affected the val performance.

From the graphs, train loss is the exact same, on track, yet val has a sudden drop in ppl and loss. The Data bugfix was important and improved performance, but I think should have affected the train values. Maybe were observing some form of grokking?

@jzhang38 do you have any reasons for this?

**Comment by jzhang38** [comment]
We sample 100 batches of val data for the actual validation.
I believe this is caused by the fact that a different partition of val data gets sampled after the dataloader fix.

**Comment by psinger** [comment]
Just different sampling should not drop the loss by that much. I believe this needs a more thorough investigation why this is happening.

**Comment by segyges** [comment]
If dataset processing was repeated between the runs it's possible the train/val split is different; if any of the data in the validation set was previously in the training set it would account for the sharp discontinuity more neatly than a different sample of val (where a random subset of 100 should really be enough that any sample delivers almost the same number).

If this is correct it doesn't mean that the val_ppl's trajectory is a bad metric, but it does mean that it isn't truly a validation set any more.

**Comment by VatsaDev** [comment]
Looks like its been fixed with train run 2? was the previous val data something it might have trained on?

**Comment by eminorhan** [comment]
It also seems odd that the fix didn't seem to have any effect on `train_loss`. Allegedly, 35% (or perhaps even 60%) of the training data was not being used previously. How can introducing that much new data not have any effect on `train_loss`, especially early after the restart? That's a huge distribution shift.

**Comment by VatsaDev** [comment]
@eminorhan I had similar thoughts, but its also possible that in 35% of new data, slipajama might be more uniform, or it could have been code data, which also has uniformity. Its strange that the Val PPL still dropping with the extra 30%, yet benchmarks are growing consistent to previous versions. is the val check contaminated?
